src/eos-updater-poll-lan.c

```diff
@@ -0,0 +1,920 @@
+
+static DlTimeCheckResult
+check_dl_time (LanData *lan_data,
+               const gchar *dl_time,
+               GDateTime **utc_out)
+{
+  gint64 utc_time;
+  gchar *utc_str_end = NULL;
+  g_autoptr(GDateTime) utc = NULL;
+
+  if (dl_time[0] == '\0')
+    return DL_TIME_INVALID;
+  errno = 0;
+  utc_time = g_ascii_strtoll (dl_time, &utc_str_end, 10);
+  if (errno == EINVAL)
+    return DL_TIME_INVALID;
+  if (errno == ERANGE)
+    return DL_TIME_OUT_OF_RANGE;
+  if (*utc_str_end != '\0')
+    return DL_TIME_INVALID;
+  utc = g_date_time_new_from_unix_utc (utc_time);
+  if (utc == NULL)
+    return DL_TIME_INVALID;
+  *utc_out = g_steal_pointer (&utc);
+  if (g_date_time_compare (lan_data->fetch_data->data->branch_file->download_time, *utc_out) >= 0)
+    return DL_TIME_OLD;
+  return DL_TIME_NEW;
+}
+
+static gboolean
+txt_v1_handler_risky_checks (LanData *lan_data,
+                             GHashTable *records,
+                             gboolean *valid,
+                             GError **error)
+{
+  if (!check_ostree_path (lan_data,
+                          g_hash_table_lookup (records, eos_avahi_v1_ostree_path ()),
+                          valid,
+                          error))
+    return FALSE;
+  if (!*valid)
+    // TODO: message
+    return TRUE;
+  return TRUE;
+}
+
+static ServiceValidResult
+txt_v1_handler_time_check (LanData *lan_data,
+                           EosServiceWithBranchFile *swbf,
+                           GHashTable *records)
+{
+  const gchar *dl_time;
+  g_autoptr(GDateTime) txt_utc = NULL;
+
+  dl_time = g_hash_table_lookup (records, eos_avahi_v1_branch_file_dl_time ());
+  switch (check_dl_time (lan_data, dl_time, &txt_utc))
+    {
+    case DL_TIME_OLD:
+      eos_service_with_branch_file_get_from_lan_data (swbf, lan_data);
+      return SERVICE_VALID;
+
+    case DL_TIME_NEW:
+      swbf->declared_download_time = g_steal_pointer (&txt_utc);
+      return SERVICE_VALID_MORE_CHECKS;
+
+    case DL_TIME_INVALID:
+      // TODO: message
+      return SERVICE_INVALID;
+
+    case DL_TIME_OUT_OF_RANGE:
+      // TODO: message
+      return SERVICE_INVALID;
+    }
+  // TODO: message
+  g_assert_not_reached ();
+  return SERVICE_INVALID;
+}
+
+static gboolean
+txt_v1_handler_checks (LanData *lan_data,
+                       EosServiceWithBranchFile *swbf,
+                       GHashTable *records)
+{
+  ServiceValidResult result;
+
+  result = txt_v1_handler_time_check (lan_data, swbf, records);
+  if (result == SERVICE_VALID_MORE_CHECKS)
+    {
+      const gchar *sha512sum;
+      const gchar *sha512sum_local;
+
+      sha512sum = g_hash_table_lookup (records, eos_avahi_v1_branch_file_sha512sum ());
+      sha512sum_local = lan_data->fetch_data->data->branch_file->contents_sha512sum;
+      if (g_strcmp0 (sha512sum_local, sha512sum) == 0)
+        eos_service_with_branch_file_get_from_lan_data (swbf, lan_data);
+      else
+        swbf->declared_sha512sum = g_strdup (sha512sum);
```

Comment:
We are checking time stamps and checksum to avoid downloading the branch file if we already have the same one.

If the checksum is different then we will maybe download it from the LAN server then.